# Pandas Skills

Few challenges that we will approach to solve using pandas and learning some tricks on the way!

in this practice I will be using a dataset from
an open-source website (kaggle.com), this dataset will be uploaded on my Github as well


In [5]:
# importing libraries 

import numpy as np
import pandas as pd


In [6]:
# reading the file dataset we will be dealing with
# check out the type of your file first
# our dataset this time is a CSV file.

slr = pd.read_csv('Salaries.csv')

In [23]:
slr.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

In [8]:
# we will try to get more info about our dataset
# using .info() method

slr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


Here we notice that we have 12 columns, and their indexes, dtype in each column 

next, we can check the average of different important columns. that will help us read and understand our dataset and see it more clearly.

In [9]:


# BasePay

slr['BasePay'].mean()

66325.44884050643

In [10]:
#OvertimePay

slr['OvertimePay'].mean()

5066.059886444622

In [11]:
#OtherPay

slr['OtherPay'].mean()

3648.7672968046095

In [12]:
slr['OtherPay'].max()

400184.25

In [13]:
slr['OtherPay'].min()

-7058.59

In [14]:
#Benefits

slr['Benefits'].mean()

25007.893150830107

In [15]:
slr['Benefits'].max()

96570.66

In [16]:
slr['Benefits'].min()

-33.89

In [17]:
#TotalPay

slr['TotalPay'].mean()

74768.321971703

In [18]:
slr['TotalPay'].max()

567595.43

In [19]:
slr['TotalPay'].min()

-618.13

In [20]:
#TotalPayBenefits

slr['TotalPayBenefits'].mean()

93692.55481056834

Now let's find out about an employee named JAMES DUDLEY, what position he is occupying, and how much he makes 

In [24]:
slr[slr['EmployeeName']== 'JAMES DUDLEY']

Id  EmployeeName                              JobTitle   BasePay  \
48  49  JAMES DUDLEY  DEPUTY CHIEF III (POLICE DEPARTMENT)  233867.6   

    OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  Notes  \
48          0.0  14851.04       NaN  248718.64         248718.64  2011    NaN   

           Agency  Status  
48  San Francisco     NaN

In [28]:
# longer virsion would be like

slr[slr['EmployeeName']=='JAMES DUDLEY']['TotalPayBenefits']

48    248718.64
Name: TotalPayBenefits, dtype: float64

#### let's grab the name of the least-paid person

In [31]:
slr [slr['TotalPayBenefits'] == slr['TotalPayBenefits'].min()]['EmployeeName']

148653    Joe Lopez
Name: EmployeeName, dtype: object

#### let's grab the name of the most paid person, in another method

In [34]:
slr.iloc[slr['TotalPayBenefits'].argmax()]

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                  167411.18
OvertimePay                                                    0.0
OtherPay                                                 400184.25
Benefits                                                       NaN
TotalPay                                                 567595.43
TotalPayBenefits                                         567595.43
Year                                                          2011
Notes                                                          NaN
Agency                                               San Francisco
Status                                                         NaN
Name: 0, dtype: object

In [35]:
# we can use the same method to find the least paid 

slr.iloc[slr['TotalPayBenefits'].argmin()]

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                    0.0
OvertimePay                                0.0
OtherPay                               -618.13
Benefits                                   0.0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Notes                                      NaN
Agency                           San Francisco
Status                                     NaN
Name: 148653, dtype: object

#### let's check the average Basepay, TotalPayBenefits for all employees per year

In [36]:
slr.groupby('Year').mean()['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [37]:
slr.groupby('Year').mean()['TotalPayBenefits']

Year
2011     71744.103871
2012    100553.229232
2013    101440.519714
2014    100250.918884
Name: TotalPayBenefits, dtype: float64

In [45]:
# to find unique job titles 

slr['JobTitle'].unique()

array(['GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY',
       'CAPTAIN III (POLICE DEPARTMENT)',
       'WIRE ROPE CABLE MAINTENANCE MECHANIC', ..., 'Conversion',
       'Cashier 3', 'Not provided'], dtype=object)

In [43]:
# to find the number of unique job titles 

slr['JobTitle'].nunique()

2159

In [46]:
# to find most 5 common jobs

slr['JobTitle'].value_counts().head()

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

In [48]:
# to find how what are job positions in the year 2013, and how many employees were in each.

slr[slr['Year']==2013]['JobTitle'].value_counts()

Transit Operator                  2295
Special Nurse                     1445
Registered Nurse                  1231
Public Svc Aide-Public Works       859
Custodian                          789
                                  ... 
Sr General Utility Mechanic          1
Sprv Adult Prob Ofc (SFERS)          1
Mayoral Staff XII                    1
Pr Investigator, Tax Collector       1
Senior Assessment Clerk              1
Name: JobTitle, Length: 1051, dtype: int64

Here we can notice how position like 'Special Nurse' has the count 1445, while we have only 1 'Mayoral Staff XII'.  

#### let's find out how many 'Manager' title we have in our dataset

In [49]:
# Here we will start with creating a function to unite the lowercase and uppercase titles we have 


def mngr_str(title):
    if 'manager' in title.lower().split():
        return True
    else:
        return False

now after creating this function, we will run the 'JobTitle' column into this function and see the results.

In [65]:
mngr = (slr['JobTitle'].apply(lambda x: mngr_str(x)))

In [58]:
mngr 

0         False
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [63]:
sum(mngr)

4056

#### This is a bonus task just to find out if there is any relationship between 'Year','BasePay', and 'TotalPayBenefits' throughout the correlation method

In [67]:
slr[['Year','BasePay','TotalPayBenefits']].corr()

Year   BasePay  TotalPayBenefits
Year              1.000000  0.033751          0.151947
BasePay           0.033751  1.000000          0.946595
TotalPayBenefits  0.151947  0.946595          1.000000

### - - - - The end - - - -